### PR TITLE
Wundef

### PR DESCRIPTION
**Pull request by peterbarker**
Having include files in the wrong order caused an object within the global plane object to have a different shape depending on which class was being compiled.

Adding -Wundef to the compiler options may have saved some pain.

### CODE REVIEW

**Comment by peterbarker** [comment]
@lucasdemarchi Sorry Lucas, I don't understand your request.

**Comment by peterbarker** [comment]
@lucasdemarchi Moving config.h would probably be Not a Small Thing :-)

Additionally, config.h/defines.h were only half the issues raised.

**Comment by lucasdemarchi** [comment]
> @lucasdemarchi Sorry Lucas, I don't understand your "Mail?" request.

a typo in your commit message :)

**Comment by peterbarker** [comment]
@lucasdemarchi Huh!  Thanks!  AP_SMTP is not ready to go in yet ;-)

**Comment by lucasdemarchi** [comment]
> @lucasdemarchi Moving config.h would probably be Not a Small Thing :-)

opening config.h to..... ugh. that's ugly. Ok, you convinced me.

**Comment by lucasdemarchi** [comment]
could you rebase this branch so semaphoreci also builds it?

**Comment by peterbarker** [comment]
@lucasdemarchi rebase (and spelling corrections...) done.

**Comment by jberaud** [comment]
I had seen the naming mistake in BattMonitor_Bebop, just forgot to fix it. So ok for me.

File: libraries/AP_NavEKF/AP_SmallEKF.h [hunk 1]
```diff
@@ -33,6 +33,8 @@
 
 #include <AP_Math/vectorN.h>
 
+#define MATH_CHECK_INDEXES 0
```
Comment (lucasdemarchi) [inline libraries/AP_NavEKF/AP_SmallEKF.h]: AP_Math/vectorN.h depends on this being set to change the behavior of the header. And you are defining it *after* the header is included. So later if you try to define it to 1 nothing will change.

**Comment by lucasdemarchi** [comment]
Another note: due to the header guards (or pragma once), the header is included only once per compilation unit. This means that even if you define for example MATH_CHECK_INDEXES to change the behavior of vectorN.h or another header, it's possible that you won't get that because another header you included before had that header as dependency (for example if you include AP_Math.h)

**Comment by lucasdemarchi** [comment]
I think the right way to fix this is to move these "#define 0" out of the user and into the header that provides it. So we would have something like this:

libraries/AP_Math/matrix3.h:
```
ifndef MATH_CHECK_INDEXES
#define MATH_CHECK_INDEXES 0
#endif
```

libraries/AP_Math/quaternion.h:
```
ifndef MATH_CHECK_INDEXES
#define MATH_CHECK_INDEXES 0
#endif
```

... And so on.

So if you want to enable MATH_CHECK_INDEXES in a single place what you do is to define it **before** including any header:
```
#define MATH_CHECK_INDEXES 1
```

**Comment by lucasdemarchi** [comment]
I applied the commits not needing a rework.

**Comment by peterbarker** [comment]
@lucasdemarchi I'm having trouble seeing how we can get a consistent shape of an object with what you're proposing with the #ifndefs - your comment mentioning header guards makes this problem clear.

SoloGimbalEKF is a really good example of this.  It changes shape depending on MATH_CHECK_INDEXES.  Defining MATH_CHECK_INDEXES to true in e.g. vector3.h may lead to undefined behaviour depending on include order.


> So if you want to enable MATH_CHECK_INDEXES in a single place what you do is to define it before including any header:
Assuming this is in a .cpp file - this will get you a consistent view when compiling THAT file - but other files may then have a different idea of the shape of that object.

I think MATH_CHECK_INDEXES should be defined once only - probably in the top of AP_Math.h.

**Comment by lucasdemarchi** [comment]
@peterbarker yeah, you are right. The problem is multiple places are using the same define to enable/disable different things.

I still think the only viable solution would be not to define MATH_CHECK_INDEXES, though and only allow it through the build system with -DMATH_CHECK_INDEXES=1

In other programs this is solved by having a config.h file with all the definitions like this. The config.h is generated at configure time and everyone uses it. config.h should be global and not only for part of the code base.  We still need a ton of work to fix it properly though. Maybe your solution is already good enough for now.

**Comment by lucasdemarchi** [comment]
-Wundef still needs to be added to waf build system and make sure we don't build external modules with it.

**Comment by peterbarker** [comment]
@lucasdemarchi This has been reworked as discussed.

**Comment by lucasdemarchi** [comment]
new version: #3639

**Pull request closed without merging**